### PR TITLE
test: e2e tests

### LINKED ISSUE #311: E2E tests that are testing on our zombienet network
# Summary

The e2e-tests will focus on using the api to construct certain tx's ie. `Foreign Assets`, `Assets`, `Local`, and `Liquid Pool Assets`, and then send them across the network then verify if they reached the other account. 

## Design

This design portion is up for revision and is a first iteration.

### **Tests Config**

There should be a test file that contains all configuration for each tx we are sending. I will demonstrate what the structure could look like below to give a better idea.

```js
// tests/index.ts file

export const tests = {
    foreignAssets: [
        {
            // This will declare the call to use
            test: 'createTransferTransaction',
            // This will be all the args for the above call
            args: ['<contains_the_args_in_the_correct_order_for_the_above_call>'],
            // This will be a tuple that will allow us to verify if the xcm message
            // succesfully went through on the other end
            verification: ['<dest_addr>', '<dest_chain_id>']
        },
        ...
    ]
}
```

### **Executor**

The executor will take in the tests config, and run each test. It should call the `balance` module to retrieve the current balance on the destination chain. Then after each xcm message is sent, and executed it should ensure with the verification module that the balance is now different than what the `balance` module gave us.

### **Verification**

The `verification` module should do 1 thing. Connect to a chain given some config, check the balance of a given account against a known value return a boolean. Then disconnect. 

### **Balance**

The `balance` module should do 1 thing. Connect to a chain given some config, check the balance of a given account, and return the relevant information.

### **Logger**

This should just format the results that are passed by the executor and log them in a nice way for tests.

## File structure

```txt
-- e2e-tests
    -- tests
        -- local.ts
        -- liquidPools.ts
        -- assets.ts
        -- foreignAssets.ts
        -- index.ts
    -- executor.ts
    -- verification.ts
    -- balance.ts
    -- logger.ts
```

### PR DESCRIPTION
**Pull request by bee344**
### Description
These are E2E tests that run using zombienet and the scripts located in the `.\scripts\` folder. The goal of these tests is to build asset transfers using the ATA and submit them, and then checking for the success of these transfers.

#### Layout
The files necessary to run the tests are located in the `.'e2e\` directory, with the exception of the setup scripts that are located on `.\scripts\`.

The tests are located in the `.\e2e\tests\` folder. It contains the functions for the `foreign-assets`. `assets`, `balances` and `pool-assets` tests, The tests parameters are defined in the `.\e2e\tests\index.ts` file, as a `TestGroups` implementation, which is an interface grouping `IndividualTest` implementations. Each `IndividualTest` consists of the name of the test to be run, the args to be passed to the test, and the verification data to check the success of the transfer.

The file `.\e2e\balance.ts` contains the function `balanceTracker` which handles the storing of the inital and final balances of an account for certain assetIds as an `IBalance` instance. This data is later used by the verificator to check the success of the transfer.

The file `.\e2e\verification.ts` contains the function `verification` which takes an instance of `IBalance` corresponding to the transfer's destination account, the assetIds and the amounts transferred and checks that the transferred amounts correlate with the destination account's final balance.

The file `.\e2e\logger.ts` contains functions tasked with displaying the evolution of the tests in a nice way.

The file `.\e2e\ consts.ts` contains constants that are used during the tests.

The file `.\e2e\executor.ts` handles the execution of the tests, retrieving the cli arguments that specify which particular test is being run, and retrieving the necessary data from the `.\e2e\tests\index.ts` file in order to run the test and verify its correct completion.

The tests also rely on two shell scripts, `.\e2e\zombienet.sh` and `.\e2e\tests.sh`. The former determines which OS the program is being run on, modifies the name of the zombienet executable accordingly and runs the zombienet. The latter handles the execution of the test itself, by determining which script is needed to run and running the corresponding setup script first and the corresponding test later.

#### Other changes

* Changed Signed Extension `ChargeTransactionPayment` by `ChargeAssetTxPayment` to prevent compatibility issues with latest versions.
* Added `ROCOCO_ASSET_HUB_SPEC_NAME` to `src/registry/Registry.ts` so that it is recognized as a `SystemChain`.

Closes #311

### CODE REVIEW

File: e2e-tests/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+### Setup
+
+To setup the testing environment you need to first download the `polkadot`, `polkadot-execute-worker`, `polkadot-prepare-worker` and `polkadot-parachain` from the `polkadot-sdk` [release page](https://github.com/paritytech/polkadot-sdk/releases/latest), as well as the `trappist-node` from its [release page](https://github.com/paritytech/trappist/releases/latest), and place them in the `../zombienet/bin/` folder.
+
+You also need to have the latest `zombienet` executable in the `../zombienet/` folder, which you can download from [here](https://github.com/paritytech/zombienet/releases/latest).
+ 
+### Launching zombienet
+
+To launch the zombienet, all you need to do is run the following commands:
```
Comment (TarikGul) [inline e2e-tests/README.md:15]:
```suggestion
To launch the zombienet run the following commands:
```

File: e2e-tests/README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+```bash
+$ yarn build && yarn e2e:build
+```
+Then you need to run:
```
Comment (TarikGul) [inline e2e-tests/README.md:19]:
```suggestion
Then run:
```

File: e2e-tests/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+### Launching the tests
+
+For testing, we provide 4 options:
+
+* Testing liquidity tokens transfers with the command `yarn e2e:liquidity-assets`.
+* Testing foreign assets transfers with the command `yarn e2e:foreign-assets`.
+* Testing local transferss with the command `yarn e2e:local`.
+* Testing assets transfers with the command `yarn e2e:assets`.
+
+Each of these commands will run the appropiate script to setup the basics, located in `../scripts/`, wait for it to finish setting up the testing environment, and then go through the tests indicated in the `./tests/index.ts` file for the chosen option.
```
Comment (TarikGul) [inline e2e-tests/README.md:34]:
```suggestion
Each of these commands will run the appropiate script to setup the basics, located in `../scripts/`. Wait for it to finish setting up the testing environment, and then go through the tests indicated in the `./tests/index.ts` file for the chosen option.
```

File: e2e-tests/balance.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
+// Copyright 2023 Parity Technologies (UK) Ltd.
+
+import { ApiPromise } from '@polkadot/api';
```
Comment (TarikGul) [inline e2e-tests/balance.ts:3]:
```suggestion
import type { ApiPromise } from '@polkadot/api';
```

File: e2e-tests/balance.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import type { FrameSystemAccountInfo, PalletAssetsAssetAccount } from '@polkadot/types/lookup';
+import type { Option } from '@polkadot/types-codec';
```
Comment (TarikGul) [inline e2e-tests/balance.ts:5]:
```suggestion
import type { Option } from '@polkadot/types-codec';

```

File: e2e-tests/executor.ts [hunk 6]
```diff
@@ -0,0 +1,166 @@
+// Copyright 2023 Parity Technologies (UK) Ltd.
+import { ApiPromise, WsProvider } from '@polkadot/api';
+import { Keyring } from '@polkadot/keyring';
+// import { KeyringPair } from '@polkadot/keyring/types';
```
Comment (TarikGul) [inline e2e-tests/executor.ts:4]: Remove this commented import

File: e2e-tests/executor.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+import { balanceTracker, IBalance } from './balance';
+import { KUSAMA_ASSET_HUB_WS_URL, ROCOCO_ALICE_WS_URL, TRAPPIST_WS_URL } from './consts';
+import { startProgressBar, startTestLogger, terminateProgressBar, testResultLogger, updateProgressBar } from './logger';
+import { assetTests, foreignAssetsTests, IndividualTest, liquidPoolsTests, localTests, tests } from './tests';
+import { verification } from './verification';
+
+const executor = async (testCase: string) => {
+	let originWsUrl = '';
+	let destWsUrl = '';
+
```
Comment (TarikGul) [inline e2e-tests/executor.ts:18]:
```suggestion
```

Comment (TarikGul) [reply e2e-tests/executor.ts:18]: Unnecessary space is all, small nit

File: e2e-tests/executor.ts [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+	let testData: IndividualTest[] = [];
+
+	await cryptoWaitReady();
+
+	const keyring = new Keyring({ type: 'sr25519' });
+
+	switch (testCase) {
```
Comment (TarikGul) [inline e2e-tests/executor.ts:25]: This switch case and the one below can be combined

File: e2e-tests/executor.ts [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+	let counter: number = 0;
+
+	startTestLogger(testCase);
+
+	const progressBar = startProgressBar(testData, testCase);
+
+	const results: [string, string, string, boolean][] = [];
+
+	for (const t of testData) {
```
Comment (TarikGul) [inline e2e-tests/executor.ts:65]: All of the manual arg management can be handled using either `ParseArgs`, or `Yargs`. Both libraries are good for handling cli input.

File: src/AssetTransferApi.ts [hunk 10]
```diff
@@ -800,7 +800,7 @@ export class AssetTransferApi {
 				'CheckMortality',
 				'CheckNonce',
 				'CheckWeight',
-				'ChargeTransactionPayment',
+				'ChargeAssetTxPayment',
```
Comment (TarikGul) [inline src/AssetTransferApi.ts:803]: Can we note in the description the changes here to the core code.

File: src/registry/Registry.ts [hunk 11]
```diff
@@ -250,7 +251,8 @@ export class Registry {
 		if (
 			POLKADOT_ASSET_HUB_SPEC_NAMES.includes(specName.toLowerCase()) ||
 			KUSAMA_ASSET_HUB_SPEC_NAMES.includes(specName.toLowerCase()) ||
-			WESTEND_ASSET_HUB_SPEC_NAMES.includes(specName.toLowerCase())
+			WESTEND_ASSET_HUB_SPEC_NAMES.includes(specName.toLowerCase()) ||
```
Comment (TarikGul) [inline src/registry/Registry.ts:254]: Same about this, lets note it in the description of the PR.

File: e2e-tests/executor.ts (hunk 9 above)
Comment (TarikGul) [reply e2e-tests/executor.ts:65]: But IMO, no need to switch now, what you did is fine, just pointing it out as another option to manage cli args.

**Review by TarikGul — APPROVED** [review]
Overall, amazing PR. There are a few nits below to be cleaned up. Also if we could add a brief description to the PR saying what was added changed etc.

File: e2e-tests/executor.ts (hunk 9 above)
Comment (bee344) [reply e2e-tests/executor.ts:64]: Will look into it, thanks for the suggestion

**Pull request merged**